File: examples_python/example_ValveController_qtm_timesync.py
from mobilerack_pybind_module import ValveController, QualisysClient
from time import sleep

valves = [14, 15]
max_pressure = 500
pressure = 300

cameras = [0,1,2,3,4,5,6,7]
number_of_rigidFrames = 4

vc = ValveController("192.168.0.100", valves, max_pressure)
qc = QualisysClient(number_of_rigidFrames, cameras, "6D")

sleep(1)  # hacky way to wait until data from qtm is received
_, timestamp = qc.getData6D()
vc.syncTimeStamp(timestamp//1000)  # sync the time to be that of QTM

for i in range(len(valves)):
    print(f"index:{i}\tvalve id:{valves[i]}\tpressure:{pressure}")
    vc.setSinglePressure(i, pressure)
    sleep(1)
    vc.setSinglePressure(i, 0)
    sleep(1)

File: examples/example_ValveController_qtm_timesync.cpp
#include "mobilerack-interface/QualisysClient.h"
#include "mobilerack-interface/ValveController.h"

int main(){
    const int pressure = 300;
    int valve_id;
    std::vector<int> map = {12, 13, 14, 15};

    ValveController vc{"192.168.0.100", map, 400};
    QualisysClient qc{"192.168.0.101", 0};

    srl::sleep(1); // hacky way to wait until data from QTM is received

    std::vector<Eigen::Transform<double, 3, Eigen::Affine>> frames_; // will not be read
    unsigned long long int timestamp;
    qc.getData(frames_, timestamp);
    fmt::print("time: {}", timestamp);
    fmt::print("time milli: {}", timestamp/1000);

    vc.syncTimeStamp(timestamp/1000); // sync timestamp to that of QTM

    for (int i = 0; i < map.size(); i++) {
        valve_id = map[i];
        std::cout << "actuator ID:\t" << i << "\tvalve ID:\t" << valve_id << "\tpressure\t" << pressure << std::endl;
        vc.setSinglePressure(i, pressure);
        srl::sleep(1);
        vc.setSinglePressure(i, 0);
        srl::sleep(1);
    }
    return 1;
}

